### PR TITLE
Chore: Extract config comment parsing

### PR DESCRIPTION
**Pull request by nzakas**
**What is the purpose of this pull request? (put an "X" next to item)**

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[x] Other, please explain:

Extracted config comment parsing into a separate utility.

**What changes did you make? (Give an overview)**

While digging around in `Linter`, I realized that the config comment parsing could be completely separated out, allowing us to better test config comment parsing. I just extracted the existing helper functions from `linter.js` and put them into a class. I also wrote up some basic tests based on the comments inside the code to verify everything works. At some point I'd like to go back and clean up the methods a bit to make them less tied to what's going on inside of `linter.js`, but for now this helps make `linter.js` smaller and gives us a place to test for further changes in comment parsing.

I'd really like to make `linter.js` smaller as it's over 1,000 lines right now and a bit hard to follow.

**Is there anything you'd like reviewers to focus on?**

Does this approach make sense?

### CODE REVIEW

**Comment by nzakas** [comment]
Hmm, looks like the build is breaking only in Node 11 due to a completely unrelated problem. MockFS `existsSync` seems to be throwing an error in Node 11 when the file isn't found instead of returning `false`: https://github.com/tschaub/mock-fs/issues/256

File: tests/lib/util/config-comment-parser.js [hunk 1]
```diff
@@ -0,0 +1,1908 @@
...
+    //     it("should report a violation when the report is right before the comment", () => {
+    //         const code = " /* eslint-disable */ ";
+
+    /*
+     *         linter.defineRule("checker", context => ({
+     *             Program() {
+     *                 context.report({ loc: { line: 1, column: 0 }, message: "foo" });
+     *             }
+     *         }));
+     *         const problems = linter.verify(code, { rules: { checker: "error" } });
```
Comment (not-an-aardvark) [inline tests/lib/util/config-comment-parser.js]: The refactoring LGTM, but I think we should remove these commented tests (or uncomment them if applicable).

Comment (ilyavolodin) [reply tests/lib/util/config-comment-parser.js]: Also, I'm guessing all of those tests were copied from linter.js test file, but I'm not seeing any removals in that file.

Comment (kaicataldo) [reply tests/lib/util/config-comment-parser.js]: Right - do we want to duplicate tests? I wonder if it would be clearer for the future to leave the tests in linter and just add new unit tests for config-comment-parser?

File: lib/util/config-comment-parser.js [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+/**
+ * Object to parse ESLint configuration comments inside JavaScript files.
+ * @name ConfigCommentParser
+ */
+module.exports = class ConfigCommentParser {
+
+    constructor() {
+        debug("Created ConfigCommentParser");
+
+        // empty for now
```
[the remark this answers is not in the document]
Comment (nzakas) [reply lib/util/config-comment-parser.js]: I was thinking at some point I'd add a `strict` option to remove some of the fallback behavior we have. Right now it's just a placeholder though. I can remove it if it's vexing.

File: tests/lib/util/config-comment-parser.js (hunk 1 above)
Comment (nzakas) [reply tests/lib/util/config-comment-parser.js]: I just copied those there temporarily to help me figure out the best tests to write for this module. The tests needs to stay in `linter.js` because they test the interpretation of the comments along with the parsing. I'll just remove these.
